### Import Libraries

In [3]:
import pandas as pd
from collections import OrderedDict
import numpy as np
import os

### Read Weather Data

In [4]:
weather_data_paths = os.listdir("data/weather")
weather_data_paths = [f"data/weather/{path}" for path in weather_data_paths]

In [5]:
# read dataframes
sheet_names = ["pr", "tas", "tasmax", "tasmin"]
data_frames = {sheet_name:[] for sheet_name in sheet_names}

for weather_data_path in weather_data_paths:
    for sheet_name in sheet_names:
        if sheet_name == "tas" and (("tasmax" in weather_data_path) or ("tasmin" in weather_data_path)):
            continue
        if sheet_name in weather_data_path:
            df = pd.read_excel(weather_data_path, sheet_name = sheet_name)
            data_frames[sheet_name].append(df)
        

In [6]:
data_frames["pr"][0]

code        name  1901-07  1902-07  1903-07  1904-07  1905-07  1906-07  \
0  GAB       Gabon  1790.41  1791.78  1902.33  1747.66  1815.31  1590.53   
1  GHA       Ghana  1279.34  1180.86  1051.27  1004.64  1132.05  1235.77   
2  GMB  The Gambia  1276.47   874.78  1010.16  1064.15  1279.24  1615.72   

   1907-07  1908-07  ...  2013-07  2014-07  2015-07  2016-07  2017-07  \
0  1756.26  1873.67  ...  1950.93  1868.19  1563.19  1816.60  1915.21   
1  1087.99  1236.69  ...  1102.36  1244.65  1106.44  1194.26  1134.55   
2   903.53  1130.89  ...  1038.67   955.89  1034.65  1030.34  1033.63   

   2018-07  2019-07  2020-07  2021-07  2022-07  
0  1768.14  2101.89  1817.35  1793.73  1734.09  
1  1315.19  1450.15  1232.15  1262.45  1230.61  
2   921.61   984.66  1050.48  1004.21  1361.17  

[3 rows x 124 columns]

### Combine Weather Data

In [27]:
# arrange pr data
pr_data = {"year":dict(), "country":dict(), "precipitation":dict()}

k= 0
for df in data_frames["pr"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            pr_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            pr_data["precipitation"][k] = pr_value
            pr_data["country"][k] = country
            pr_data["year"][k] = year

            k +=1

pr_data_df = pd.DataFrame.from_dict(pr_data)
pr_data_df.head()

year country  precipitation
0  1901   Gabon        1790.41
1  1902   Gabon        1791.78
2  1903   Gabon        1902.33
3  1904   Gabon        1747.66
4  1905   Gabon        1815.31

In [28]:
pr_data_df.shape

(2806, 3)

In [29]:
# arrange tas:Average Mean Surface Air Temperature data
tas_data = {"year":dict(), "country":dict(), "average_mean_surface_air_temp":dict()}

k= 0
for df in data_frames["tas"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tas_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tas_data["average_mean_surface_air_temp"][k] = tas_value
            tas_data["country"][k] = country
            tas_data["year"][k] = year

            k +=1

tas_data_df = pd.DataFrame.from_dict(tas_data)
tas_data_df.head()

year country  average_mean_surface_air_temp
0  1901   Gabon                          24.51
1  1902   Gabon                          24.71
2  1903   Gabon                          24.85
3  1904   Gabon                          24.55
4  1905   Gabon                          24.94

In [30]:
tas_data_df.shape

(2806, 3)

In [31]:
# arrange tasmax:Average Maximum Surface Air Temperature data
tasmax_data = {"year":dict(), "country":dict(), "average_maximum_surface_air_temp":dict()}

k= 0
for df in data_frames["tasmax"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tasmax_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tasmax_data["average_maximum_surface_air_temp"][k] = tasmax_value
            tasmax_data["country"][k] = country
            tasmax_data["year"][k] = year

            k +=1

tasmax_data_df = pd.DataFrame.from_dict(tasmax_data)
tasmax_data_df.head()

year                   country  average_maximum_surface_air_temp
0  1901  Central African Republic                             31.34
1  1902  Central African Republic                             31.38
2  1903  Central African Republic                             31.39
3  1904  Central African Republic                             31.36
4  1905  Central African Republic                             31.45

In [32]:
tasmax_data_df.shape

(2806, 3)

In [33]:
# arrange tasmin:Average Minimum Surface Air Temperature data
tasmin_data = {"year":dict(), "country":dict(), "average_minimum_surface_air_temp":dict()}

k= 0
for df in data_frames["tasmin"]:
    for i in range(df.shape[0]):
        for j in range(2,df.shape[1]):
            tasmin_value = df.iloc[i,j]
            country = df["name"][i]
            year = df.columns[j].split("-")[0].strip()
            tasmin_data["average_minimum_surface_air_temp"][k] = tasmin_value
            tasmin_data["country"][k] = country
            tasmin_data["year"][k] = year

            k +=1

tasmin_data_df = pd.DataFrame.from_dict(tasmin_data)
tasmin_data_df.head()

year                   country  average_minimum_surface_air_temp
0  1901  Central African Republic                             18.60
1  1902  Central African Republic                             18.65
2  1903  Central African Republic                             18.66
3  1904  Central African Republic                             18.65
4  1905  Central African Republic                             18.63

In [34]:
tasmin_data_df.shape

(2806, 3)

In [35]:
combined_weather_data = {"year":dict(), 
                 "country":dict(), 
                 "precipitation":dict(),
                 "average_mean_surface_air_temp":dict(),
                 "average_maximum_surface_air_temp":dict(),
                 "average_minimum_surface_air_temp":dict()
                }

for i in range(len(pr_data_df)):
    year = pr_data_df["year"][i]
    country = pr_data_df["country"][i]
    precipitation = pr_data_df["precipitation"][i]
    
    mean_temp_index = np.where((tas_data_df["year"] == year) & (tas_data_df["country"]==country))[0][0]
    mean_temp = tas_data_df["average_mean_surface_air_temp"][mean_temp_index]

    max_temp_index = np.where((tasmax_data_df["year"] == year) & (tasmax_data_df["country"]==country))[0][0]
    max_temp = tasmax_data_df["average_maximum_surface_air_temp"][max_temp_index]

    min_temp_index = np.where((tasmin_data_df["year"] == year) & (tasmin_data_df["country"]==country))[0][0]
    min_temp = tasmin_data_df["average_minimum_surface_air_temp"][min_temp_index]

    combined_weather_data["year"][i] = year
    combined_weather_data["country"][i] = country
    combined_weather_data["precipitation"][i] = precipitation
    combined_weather_data["average_mean_surface_air_temp"][i] = mean_temp
    combined_weather_data["average_maximum_surface_air_temp"][i] = max_temp
    combined_weather_data["average_minimum_surface_air_temp"][i] = min_temp

combined_weather_data_df = pd.DataFrame.from_dict(combined_weather_data)
combined_weather_data_df.head()

year country  precipitation  average_mean_surface_air_temp  \
0  1901   Gabon        1790.41                          24.51   
1  1902   Gabon        1791.78                          24.71   
2  1903   Gabon        1902.33                          24.85   
3  1904   Gabon        1747.66                          24.55   
4  1905   Gabon        1815.31                          24.94   

   average_maximum_surface_air_temp  average_minimum_surface_air_temp  
0                             28.70                             20.37  
1                             28.90                             20.58  
2                             29.04                             20.72  
3                             28.73                             20.41  
4                             29.13                             20.81

In [16]:
combined_weather_data_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2806 entries, 0 to 2805
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   year                              2806 non-null   object 
 1   country                           2806 non-null   object 
 2   precipitation                     2806 non-null   float64
 3   average_mean_surface_air_temp     2806 non-null   float64
 4   average_maximum_surface_air_temp  2806 non-null   float64
 5   average_minimum_surface_air_temp  2806 non-null   float64
dtypes: float64(4), object(2)
memory usage: 153.5+ KB


Apparently there are no missing values in the weather dataset

### Read Malaria Incidence Data

In [17]:
malaria_data = pd.read_csv("data/malaria-incidence.csv")
malaria_data.head()

IND_UUID       IND_NAME               IND_CODE  DIM_GEO_CODE_M49  \
0  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                 4   
1  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE               953   
2  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                24   
3  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE               954   
4  442CEA8  Malaria cases  MALARIA_EST_INCIDENCE                32   

  GEO_NAME_SHORT  DIM_TIME DIM_TIME_TYPE DIM_VALUE_TYPE  VALUE_NUMERIC  \
0    Afghanistan      2000          YEAR  RATE_PER_1000       87.15783   
1         Africa      2000          YEAR  RATE_PER_1000      372.59359   
2         Angola      2000          YEAR  RATE_PER_1000      325.70361   
3       Americas      2000          YEAR  RATE_PER_1000       14.12108   
4      Argentina      2000          YEAR  RATE_PER_1000        2.37385   

   VALUE_NUMERIC_LOWER  VALUE_NUMERIC_UPPER DIM_PUBLISH_STATE_CODE  
0             55.97768            134.25330              PUBLISHED  
1            343.83142            404.60421              PUBLISHED  
2            197.40626            506.60785              PUBLISHED  
3             12.76521             15.58999              PUBLISHED  
4              2.37385              2.37385              PUBLISHED

In [18]:
# rename "The Gambia" to "Gambia" and "São Tomé and Príncipe" to "Sao Tome and Principe"
combined_weather_data_df = combined_weather_data_df.replace(["The Gambia", "São Tomé and Príncipe"], ["Gambia", "Sao Tome and Principe"])

countries_in_malaria_data = malaria_data["GEO_NAME_SHORT"].unique()
west_central_african_countries = combined_weather_data_df["country"].unique()

In [19]:
# how many west african countries have matched malaria incidence data?
matched_countries = []
for country in west_central_african_countries:
    if country in countries_in_malaria_data:
        matched_countries.append(country)
    else:
        print(country)

print(len(matched_countries))

23


Great! All countries are matched!

### Combine both datasets
To combine both datasets, we only need to obtain the malaria incidence values for each year and each country from the malaria dataset and match it to the corresponding country and year in the weather dataset.

In [81]:
combined_weather_data_df["year"] = combined_weather_data_df["year"].astype("int")
malaria_data["DIM_TIME"] = malaria_data["DIM_TIME"].astype("int")
malaria_years = [year for year in malaria_data["DIM_TIME"].unique()]
n = combined_weather_data_df.shape[0]
m = malaria_data.shape[0]

weather_match_indices = {f'{combined_weather_data_df["country"][i]}_and_{combined_weather_data_df["year"][i]}':i for i in range(n)}
incidence_matches = [f'{malaria_data["GEO_NAME_SHORT"][i]}_and_{malaria_data["DIM_TIME"][i]}' for i in range(m)]

malaria_incidence_values = {"incidence":{}}
k=0
others = []
for i in range(combined_weather_data_df.shape[0]):
    year = combined_weather_data_df["year"][i]
    country = combined_weather_data_df["country"][i]
    if f"{country}_and_{year}" in incidence_matches:
        k +=1

        index = np.where((malaria_data["DIM_TIME"] == year) & (malaria_data["GEO_NAME_SHORT"]==country))[0][0]
        incidence = malaria_data["VALUE_NUMERIC"][index]
            
    else:
        incidence = None
    malaria_incidence_values["incidence"][i] = incidence

incidence_df = pd.DataFrame.from_dict(malaria_incidence_values)
incidence_df.head()

incidence
0        NaN
1        NaN
2        NaN
3        NaN
4        NaN

In [82]:
# combine all data
data = pd.concat([combined_weather_data_df, pd.DataFrame.from_dict(malaria_incidence_values)], axis = 1)
data = data[data["incidence"].notna()].reset_index(drop = True)
data.head()

year country  precipitation  average_mean_surface_air_temp  \
0  2000   Gabon        1951.49                          24.96   
1  2001   Gabon        1749.28                          25.11   
2  2002   Gabon        2015.69                          25.30   
3  2003   Gabon        1856.99                          25.27   
4  2004   Gabon        1725.16                          25.26   

   average_maximum_surface_air_temp  average_minimum_surface_air_temp  \
0                             29.15                             20.83   
1                             29.30                             20.98   
2                             29.49                             21.17   
3                             29.46                             21.14   
4                             29.44                             21.12   

   incidence  
0  330.43243  
1  322.67046  
2  293.25570  
3  248.70030  
4  187.82734

In [83]:
data.shape

(462, 7)

The combined dataset is made up of 462 entries which are enough for modeling.

In [85]:
# save data
data.to_csv("data/processed_data.csv", index = False)